scheduler/reconcile.go

```diff
@@ -1192,19 +1193,33 @@ func (a *allocReconciler) reconcileReconnecting(reconnecting allocSet, all alloc
 			continue
 		}
 
+		// A replacement allocation could fail and be replaced with another
+		// so follow the replacements in a linked list style
+		replacements := []string{}
+		nextAlloc := reconnectingAlloc.NextAllocation
+		for {
+			val, ok := all[nextAlloc]
+			if !ok {
+				break
+			}
+			replacements = append(replacements, val.ID)
+			nextAlloc = val.NextAllocation
+		}
+
 		// Find replacement allocations and decide which one to stop. A
 		// reconnecting allocation may have multiple replacements.
 		for _, replacementAlloc := range all {
 
-			// Skip allocations that are not a replacement of the one
-			// reconnecting.
-			isReplacement := replacementAlloc.ID == reconnectingAlloc.NextAllocation
+			// Skip the allocation if it is the reconnecting alloc
+			if replacementAlloc == reconnectingAlloc {
+				continue
+			}
 
-			// Skip allocations that are server terminal.
+			// Skip allocations that are server terminal or not replacements.
 			// We don't want to replace a reconnecting allocation with one that
 			// is or will terminate and we don't need to stop them since they
 			// are already marked as terminal by the servers.
-			if !isReplacement || replacementAlloc.ServerTerminalStatus() {
+			if !slices.Contains(replacements, replacementAlloc.ID) || replacementAlloc.ServerTerminalStatus() {
```

Comment:
Should we check both client and server terminal status? If the allocation is client-terminal I don't think we can use it as our choice, because it's stopped already.

```suggestion
			if !slices.Contains(replacements, replacementAlloc.ID) || replacementAlloc.TerminalStatus() {
```

Reply:
@Juanadelacuesta and I discussed this today.  We _can_ but then have to do another check to see if it's client terminal, and if so, stop the alloc.  It's a bit redundant because this scenario is caught when we decide which alloc to keep.